File: src/main/java/net/dblsaiko/illuminate/client/api/Light.java
package net.dblsaiko.illuminate.client.api;

import net.minecraft.util.Identifier;
import org.jetbrains.annotations.NotNull;
import org.joml.Vector3fc;

/**
 * An oriented light which projects a texture.
 */
public interface Light {
    /**
     * Identifier for the texture that should be projected.
     */
    @NotNull
    Identifier tex();

    /**
     * Returns the origin point of the light.
     */
    @NotNull
    Vector3fc pos();

    /**
     * Returns the yaw rotation of the light.
     */
    float yaw();

    /**
     * Returns the pitch rotation of the light.
     */
    default float pitch() {
        return 0f;
    }

    /**
     * Returns the roll rotation of the light.
     */
    default float roll() {
        return 0f;
    }

    /**
     * Returns the vertical field of view (FoV) of the light.
     */
    default float fov() {
        return 90f;
    }

    /**
     * Returns the aspect ratio of the light (X:Y).
     */
    default float aspect() {
        return 1f;
    }

    /**
     * Returns the distance of the near clip plane of the projection from the origin.
     */
    default float near() {
        return 0.8660254f;
    }

    /**
     * Returns the distance of the far clip plane of the projection from the origin.
     */
    default float far() {
        return 50f;
    }

    /**
     * Prepare the light for rendering. This gets called before the world gets drawn from the light's perspective.
     */
    default void prepare(float delta) {
    }
}

File: src/main/kotlin/client/api/Light.kt
package therealfarfetchd.illuminate.client.api

import org.joml.Vector3f

/**
 * An oriented light which projects a texture.
 */
interface Light {

  /**
   * A GL texture id representing the texture that should be projected.
   */
  val tex: Int

  /**
   * The origin point of the light
   */
  val pos: Vector3f

  /**
   * The yaw rotation of the light
   */
  val yaw: Float

  /**
   * The pitch rotation of the light
   */
  @JvmDefault
  val pitch: Float
    get() = 0f

  /**
   * The roll rotation of the light
   */
  @JvmDefault
  val roll: Float
    get() = 0f

  /**
   * The vertical field of view (FoV) of the light
   */
  @JvmDefault
  val fov: Float
    get() = 90f

  /**
   * The aspect ratio of the light (X:Y)
   */
  @JvmDefault
  val aspect: Float
    get() = 1f

  /**
   * The distance of the near clip plane of the projection from the origin
   */
  @JvmDefault
  val near: Float
    get() = 0.8660254f

  /**
   * The distance of the far clip plane of the projection from the origin
   */
  @JvmDefault
  val far: Float
    get() = 50f

  /**
   * Prepare the light for rendering. This gets called before the world gets drawn from the light's perspective
   */
  @JvmDefault
  fun prepare(delta: Float) {
  }

}
